File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
function solution(bandage, health, attacks) {
  const prepareTime = bandage[0];
  const recoveryPerSecond = bandage[1];
  const additionalRecovery = bandage[2];

  let answer = health;
  for (let i = 0; i < attacks.length; i++) {
    answer -= attacks[i][1];

    if (answer <= 0) return -1;

    if (i !== attacks.length - 1) {
      const time = attacks[i + 1][0] - attacks[i][0] - 1;
      const additional = Math.floor(time / prepareTime);
      answer += time * recoveryPerSecond + additionalRecovery * additional;
      if (answer >= health) answer = health;
    }
  }

  return answer <= 0 ? -1 : answer;
}

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
from typing import List

def solution(bandage: List[int], health: int, attacks: List[List[int]]) -> int:
    cast_time, heal_per_second, bonus_heal = bandage
    current = health
    for idx, (time, damage) in enumerate(attacks):
        current -= damage
        if current <= 0:
            return -1
        if idx == len(attacks) - 1:
            break
        next_time = attacks[idx + 1][0]
        gap = next_time - time - 1
        if gap <= 0:
            continue
        full_casts, _ = divmod(gap, cast_time)
        current += gap * heal_per_second + full_casts * bonus_heal
        if current > health:
            current = health
    return -1 if current <= 0 else current
